extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/tenant/HibernateMultiTenantConnectionProvider.java

```diff
@@ -0,0 +1,78 @@
+package io.quarkus.hibernate.orm.runtime.tenant;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.hibernate.engine.jdbc.connections.spi.AbstractMultiTenantConnectionProvider;
+import org.hibernate.engine.jdbc.connections.spi.ConnectionProvider;
+import org.jboss.logging.Logger;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.InstanceHandle;
+
+/**
+ * Maps from the Quarkus {@link TenantConnectionResolver} to the {@link HibernateMultiTenantConnectionProvider} model.
+ * 
+ * @author Michael Schnell
+ *
+ */
+public class HibernateMultiTenantConnectionProvider extends AbstractMultiTenantConnectionProvider {
+
+    private static final long serialVersionUID = 1L;
+
+    private static final Logger LOG = Logger.getLogger(HibernateMultiTenantConnectionProvider.class);
+
+    private static final Map<String, ConnectionProvider> PROVIDER_MAP = new ConcurrentHashMap<>();
+
+    @Override
+    protected ConnectionProvider getAnyConnectionProvider() {
+        String tenantId = tenantResolver().getDefaultTenantId();
+        if (tenantId == null) {
+            throw new IllegalStateException("Method 'TenantResolver.getDefaultTenantId()' returned a null value. "
+                    + "This violates the contract of the interface!");
+        }
+        return selectConnectionProvider(tenantId);
+    }
+
+    @Override
+    protected ConnectionProvider selectConnectionProvider(String tenantIdentifier) {
+        LOG.debugv("selectConnectionProvider({0})", tenantIdentifier);
+
+        return PROVIDER_MAP.computeIfAbsent(tenantIdentifier, tid -> {
```

Comment:
I'd prefer if you could use an "old fashioned get" here first, only falling back to a `computeIfAbsent` operation if your local cache doesn't have it so to avoid contention in the implementation of `ConcurrentHashMap#computeIfAbsent` since there is no need to have contention.

Reply:
OK, I changed it